src/main/java/io/opentracing/contrib/global/GlobalTracer.java

```diff
@@ -0,0 +1,148 @@
+ * <li>Automatically by using the Java <code>ServiceLoader</code> SPI mechanism to load an implementation
+ * from the classpath.</li>
+ * </ol>
+ */
+public final class GlobalTracer implements Tracer {
+    private static final Logger LOGGER = Logger.getLogger(GlobalTracer.class.getName());
+
+    private static final GlobalTracer INSTANCE = new GlobalTracer();
+
+    /**
+     * The resolved {@link Tracer} to delegate the global tracing implementation to.<br>
+     * This can be either an {@link #set(Tracer) explicitly set delegate}
+     * or the automatically resolved Tracer implementation.
+     */
+    private final AtomicReference<Tracer> globalTracer = new AtomicReference<Tracer>();
+
+    private GlobalTracer() {
+    }
+
+    private Tracer lazyTracer() {
+        Tracer instance = globalTracer.get();
+        if (instance == null) {
+            final Tracer singleton = loadSingleton();
+            while (instance == null && singleton != null) { // for race condition only, should rarely happen.
+                globalTracer.compareAndSet(null, singleton);
+                instance = globalTracer.get();
+            }
+            LOGGER.log(Level.INFO, "Using GlobalTracer implementation: {0}.", instance);
+        }
+        LOGGER.log(Level.FINEST, "GlobalTracer: {0}.", instance);
+        return instance;
+    }
+
+    /**
+     * Returns the {@linkplain GlobalTracer} instance.
+     * Upon first use of any tracing method, this tracer lazily determines which actual {@link Tracer}
+     * implementation to use:
+     * <ol type="a">
+     * <li>If an explicitly configured tracer was provided via the {@link #set(Tracer)} method,
+     * that will always take precedence over automatically provided tracer instances.</li>
+     * <li>A Tracer implementation can be automatically provided using the Java {@link ServiceLoader} through the
+     * <code>META-INF/services/io.opentracing.Tracer</code> service definition file.<br>
+     * The {@linkplain GlobalTracer} will not attempt to choose between implementations;
+     * if more than one is found by the {@linkplain ServiceLoader service loader},
+     * a warning is logged and tracing is disabled by falling back to the default implementation:</li>
+     * <li>If no tracer implementation is available, the {@link io.opentracing.NoopTracer NoopTracer}
+     * will be used.</li>
+     * </ol>
+     *
+     * @return The global tracer.
+     */
+    public static Tracer get() {
+        return INSTANCE;
+    }
+
+    /**
+     * Explicit registration of a configured {@link Tracer} to back the behaviour
+     * of the {@link #get() global tracer}.
+     * <p>
+     * The previous global tracer is returned so it can be restored later if necessary.
+     *
+     * @param tracer Tracer to use as global tracer.
+     * @return The previous global tracer.
+     */
+    public static Tracer set(final Tracer tracer) {
```

Comment:
@yurishkuro agrees with you on the simpler design, to which I replied the following:

> While I don't recommend using GlobalTracer.get() as a constant anywhere, this is a library and I cannot prohibit using it this way. Especially since tracing is similar to logging in my mind and that's how people are already used to working with loggers.
> This way, the behaviour of the global tracer is known before and after application initialization for all references (obtained either before or after initialization completed). Note due to the atomic reference, for observer of GlobalTracer there is no 'during initialization' which would make things really unpredictable.
> A second advantage is that it forces people to stick to the API and not take implementation-specific shortcuts, because that's all they get from the singleton tracer.
> Thirdly, it saves me from explaining why the GlobalTracer is not a Tracer, as you correctly mentioned previously 😉 .

I agree symmetrical design feels simpler, but would potentially lead to unexpected behaviour when people use the global tracer as a constant, resulting in unpredictable functionality if the initialization order may vary depending on timing issues (worst-case some constants resolving to no-op tracer, some to a auto-registered Tracer service and another to a (later-) initialized explicit set() implementation.

I wouldn't call this _resolved_ yet, as I have not received feedback to these points, which I agree are just my opinion.